api/metric/api.go

```diff
@@ -18,64 +18,169 @@ import (
 	"context"
 
 	"go.opentelemetry.io/api/core"
-	"go.opentelemetry.io/api/registry"
 	"go.opentelemetry.io/api/unit"
 )
 
-type MetricType int
+// Kind categorizes different kinds of metric.
+type Kind int
 
+//go:generate stringer -type=Kind
 const (
-	Invalid    MetricType = iota
-	Gauge                 // Supports Set()
-	Cumulative            // Supports Inc()
+	Invalid        Kind = iota
+	CumulativeKind      // Supports Add()
+	GaugeKind           // Supports Set()
+	MeasureKind         // Supports Record()
 )
 
+// Recorder is the implementation-level interface to Set/Add/Record individual metrics.
+type Recorder interface {
+	// Record allows the SDK to observe a single metric event
+	Record(ctx context.Context, value float64)
+}
+
+// LabelSet represents a []core.KeyValue for use as pre-defined labels
+// in the metrics API.
+//
+// TODO this belongs outside the metrics API, in some sense, but that
+// might create a dependency.  Putting this here means we can't re-use
+// a LabelSet between metrics and tracing, even when they are the same
+// SDK.
+type LabelSet interface {
+	Meter() Meter
+}
+
+// Meter is an interface to the metrics portion of the OpenTelemetry SDK.
 type Meter interface {
-	// TODO more Metric types
-	GetFloat64Gauge(ctx context.Context, gauge *Float64GaugeHandle, labels ...core.KeyValue) Float64Gauge
+	// DefineLabels returns a reference to a set of labels that
+	// cannot be read by the application.
+	DefineLabels(context.Context, ...core.KeyValue) LabelSet
+
+	// RecorderFor returns a handle for observing single measurements.
+	RecorderFor(context.Context, LabelSet, Instrument) Recorder
+
+	// RecordSingle records a single measurement without computing a handle.
+	RecordSingle(context.Context, LabelSet, Measurement)
+
+	// RecordBatch atomically records a batch of measurements.  An
+	// implementation may elect to call `RecordSingle` on each
+	// measurement, or it could choose a more-optimized approach.
+	RecordBatch(context.Context, LabelSet, ...Measurement)
 }
 
-type Float64Gauge interface {
-	Set(ctx context.Context, value float64, labels ...core.KeyValue)
+type InstrumentID uint64
+
+// Instrument represents a named metric with recommended local-aggregation keys.
+type Instrument struct {
```

Comment:
this is more like what we called in the protocol and what OpenMetrics as well calls "MetricDescriptor"

Reply:
I'm open to a change of `Instrument` to `Descriptor`.